DependencyGraphBuilder and log4j-api 2

StaticLinkageChecker does not work with Log4j 2 because it cannot fetch zip file for log4j-api-java9.

`org.apache.logging.log4j:log4j-api:2.11.1` cannot be resolved by `DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph`, because its dependency to org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 with scope: provided.

```
org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)

	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:355)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.resolveCompileTimeDependencies(DependencyGraphBuilder.java:132)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph(DependencyGraphBuilder.java:165)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilderTest.testGetDirectDependencies_missingDependency(DependencyGraphBuilderTest.java:99)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:422)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:224)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:338)
	... 25 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:48)
```

```
    <dependency>
      <groupId>org.apache.logging.log4j</groupId>
      <artifactId>log4j-api-java9</artifactId>
      <scope>provided</scope>
      <type>zip</type>
    </dependency>
```

The log4j-api-java9 artifact is not available in Maven central.

`DependencyGraphBuilder.getCompleteDependencies` and `getTransitiveDependencies` are fine because they don't try to fetch dependencies with scope: provided.

log4j-api-java9 has mostly dummy classes to allow module-info.java to compile.

![image](https://user-images.githubusercontent.com/28604/51189999-3e97f400-18af-11e9-93c2-f51191c2e154.png)

# log4j-core-java9

This also has similar structure. log4j-core depends on log4j-core-java9 with zip and provided.

https://search.maven.org/artifact/org.apache.logging.log4j/log4j-core/2.11.1/jar

log4j-core-java9/src/assembly/java9.xml only includes `module-info.class` to output directory


# log4j-api-java9

Let's build https://github.com/apache/logging-log4j2

Log4j uses Maven toolchain plugin

```
~$ cat ~/.m2/toolchains.xml 
<?xml version="1.0" encoding="UTF8"?>
<toolchains>
  <!-- JDK toolchains -->
  <toolchain>
    <type>jdk</type>
    <provides>
      <version>1.8</version>
      <vendor>openjdk</vendor>
    </provides>
    <configuration>
      <jdkHome>/usr/local/buildtools/java/jdk</jdkHome>
    </configuration>
  </toolchain>
  <toolchain>
    <type>jdk</type>
    <provides>
      <version>9</version>
      <vendor>openjdk</vendor>
    </provides>
    <configuration>
      <jdkHome>/usr/local/buildtools/java/jdk9</jdkHome>
    </configuration>
  </toolchain>
</toolchains>
```


To understand the purpose of log4j-api-java9.


`mvn install -DskipTests` installed the zip to m2 directory:

```
$ find ~/.m2 -name 'log4j-api-java9*'
/usr/local/google/home/suztomo/.m2/repository/org/apache/logging/log4j/log4j-api-java9
/usr/local/google/home/suztomo/.m2/repository/org/apache/logging/log4j/log4j-api-java9/3.0.0-SNAPSHOT/log4j-api-java9-3.0.0-SNAPSHOT.pom
/usr/local/google/home/suztomo/.m2/repository/org/apache/logging/log4j/log4j-api-java9/3.0.0-SNAPSHOT/log4j-api-java9-3.0.0-SNAPSHOT.zip
```

log4j-api-java9/src/assembly/java9.xml only includes `module-info.class`, `DefaultObjectInputFilter.class`, `StackLocator.class`, `ProcessIdUtil.class` to its output directory.

These classes are then included in log4j-api.jar under META-INF/versions/9.

```
~$ jar tf /usr/local/google/home/suztomo/.m2/repository/org/apache/logging/log4j/log4j-api/3.0.0-SNAPSHOT/log4j-api-3.0.0-SNAPSHOT.jar |grep 'versions/9'
META-INF/versions/9/
META-INF/versions/9/org/
META-INF/versions/9/org/apache/
META-INF/versions/9/org/apache/logging/
META-INF/versions/9/org/apache/logging/log4j/
META-INF/versions/9/org/apache/logging/log4j/util/
META-INF/versions/9/org/apache/logging/log4j/util/internal/
META-INF/versions/9/org/apache/logging/log4j/util/ProcessIdUtil.class
META-INF/versions/9/org/apache/logging/log4j/util/StackLocator.class
META-INF/versions/9/module-info.class
META-INF/versions/9/org/apache/logging/log4j/util/internal/DefaultObjectInputFilter.class
```

Does this work without log4j-api/pom.xml depending on log4j-api-java9?

Yes, it does. Then what's the point of having the dependency in pom.xml?

### META-INF/versions/9

`log4j-api-java9`'s `assembly/java9.xml` defines classes are to be in `/classes/META-INF/versions/9`. 

### A release note mentioning Java 9

The release note on 2.10 says:

> The Log4j API jar is now a multi-release jar to provide implementations of Java 9 specific classes. 

https://github.com/apache/logging-log4j2/blob/master/RELEASE-NOTES.md#apache-log4j-2100-release-notes

## Log4j compatibility with Java 7 and 8

Log4j's documentation says it's compatible with Java 7. `Javap` command verifies it:

```
$ javap -cp `find ~/.m2 -name "log4j-api-2.11.1.jar"` -v org.apache.logging.log4j.Logger |head
Classfile jar:file:/usr/local/google/home/suztomo/.m2/repository/org/apache/logging/log4j/log4j-api/2.11.1/log4j-api-2.11.1.jar!/org/apache/logging/log4j/Logger.class
  Last modified Jul 22, 2018; size 17739 bytes
  MD5 checksum d666a51086b74bc37ed70bb8ea012983
  Compiled from "Logger.java"
public interface org.apache.logging.log4j.Logger
  minor version: 0
  major version: 51
  flags: ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT
Constant pool:
    #1 = Class              #173          // org/apache/logging/log4j/Logger

```

In github master, log4j's root pom.xml has the following lines to make it compatible with Java 8.

```
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
```

Javap confirms the version 52 (Java 8)

```
$ javap -cp log4j-api/target/classes -v org/apache/logging/log4j/Logger |head
Classfile /usr/local/google/home/suztomo/logging-log4j2/log4j-api/target/classes/org/apache/logging/log4j/Logger.class
  Last modified Jan 15, 2019; size 17539 bytes
  MD5 checksum 07f0725b56dc60537ceeca9b22722aa3
  Compiled from "Logger.java"
public interface org.apache.logging.log4j.Logger
  minor version: 0
  major version: 52
  flags: ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT
Constant pool:
    #1 = Class              #167          // org/apache/logging/log4j/Logger
```

Log4j developers aim to relesae version 3 to support Java 8 and higher. ([[log4j] Toward 3.0](https://lists.apache.org/thread.html/9b05d3cd7d768a6ef590f03e9616fdd33bc7c0ff8e4b372bdb048bb1@%3Cdev.logging.apache.org%3E))




These values are overwritten at log4j-api-java9/pom.xml. The classes in log4j-api-java9 are of course compiled for Java 9:

```
~/logging-log4j2$ /usr/local/buildtools/java/jdk9/bin/javap -cp log4j-api-java9/target/classes  -v org.apache.logging.log4j.util.ProcessIdUtil  |head
Classfile /usr/local/google/home/suztomo/logging-log4j2/log4j-api-java9/target/classes/org/apache/logging/log4j/util/ProcessIdUtil.class
  Last modified Jan 15, 2019; size 778 bytes
  MD5 checksum 40f9795a6895bef690ab32e25a1085bd
  Compiled from "ProcessIdUtil.java"
public class org.apache.logging.log4j.util.ProcessIdUtil
  minor version: 0
  major version: 53
  flags: (0x0021) ACC_PUBLIC, ACC_SUPER
  this_class: #6                          // org/apache/logging/log4j/util/ProcessIdUtil
  super_class: #8                         // java/lang/Object

```

Classes in log4j-api-java9 have ones with same names in log4j-core, but the implementation in log4j-api-java9 module is optimized by leveraging Java9's new classes. Example of `ProcessIdUtil`:

![image](https://user-images.githubusercontent.com/28604/51201928-6006d900-18cb-11e9-8eae-f0cc54e3cfd6.png)

Log4j-api-java9 also includes module-info.java for new module system, Java9's biggest change.

### maven-dependency-plugin:unpack-classes

log4j-api/pom.xml unpacks class files from log4j-api-java9 into its class directory

```
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-dependency-plugin</artifactId>
        <version>3.0.2</version>
        <executions>
          <execution>
            <id>unpack-classes</id>
            <phase>prepare-package</phase>
            <goals>
              <goal>unpack</goal>
            </goals>
            <configuration>
              <artifactItems>
                <artifactItem>
                  <groupId>org.apache.logging.log4j</groupId>
                  <artifactId>log4j-api-java9</artifactId>
                  <version>${project.version}</version>
                  <type>zip</type>
                  <overWrite>false</overWrite>
                </artifactItem>
              </artifactItems>
              <includes>**/*.class</includes>
              <excludes>**/*.java</excludes>
              <outputDirectory>${project.build.directory}</outputDirectory>
              <overWriteReleases>false</overWriteReleases>
              <overWriteSnapshots>true</overWriteSnapshots>
            </configuration>
          </execution>
        </executions>
      </plugin>
```

```
$ find log4j-api -name 'module-info.class'
log4j-api/target/classes/META-INF/versions/9/module-info.class
```

### Multi-Release JAR Files

[JEP 238](http://openjdk.java.net/jeps/238) describes the purpose of `META-INF/versions/9` directory inside a jar file.

> Suppose there are alternate versions of A and B that can take advantage of Java 9 features. We can bundle them into a single JAR as follows:
>
> jar root
>  - A.class
>  - B.class
>  - C.class
>  - D.class
>  - META-INF
>     - versions
>        - 9
>           - A.class
>           - B.class

# Does log4j-api work without the provided dependency?

Yes, it does. After commenting out the dependency to log4j-api-java9 from log4j-api, the build peocess still worked:

```
[INFO] --- maven-dependency-plugin:3.0.2:unpack (unpack-classes) @ log4j-api ---
Downloading: 
...
[INFO] Configured Artifact: org.apache.logging.log4j:log4j-api-java9:3.0.0-SNAPSHOT:zip
[INFO] Unpacking /usr/local/google/home/suztomo/logging-log4j2/log4j-api-java9/target/log4j-api-java9-3.0.0-SNAPSHOT.zip to /usr/local/google/home/suztomo/logging-log4j2/log4j-api/target with includes "**/*.class" and excludes "**/*.java"
[INFO] 
...

$ find log4j-api/target/classes -name 'module-info.class'
log4j-api/target/classes/META-INF/versions/9/module-info.class
```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -161,9 +161,25 @@ class ClassDumper {
 
     SymbolReferenceSet.Builder symbolTableBuilder = SymbolReferenceSet.builder();
     for (JavaClass javaClass : listClassesInJar(jarFilePath)) {
+      if (!isCompatibleClassFileVersion(javaClass)) {
+        continue;
+      }
       symbolTableBuilder.addAll(scanSymbolReferencesInClass(javaClass));
     }
     return symbolTableBuilder.build();
+  }
+
+  /**
+   * Returns true if {@code javaClass} file format is compatible with this tool. As of January 2019,
+   * Java 8 is supported.
+   *
+   * @see <a href="https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1">Java
+   *     Virtual Machine Specification: The ClassFile Structure: minor_version, major_version</a>
+   */
+  private static boolean isCompatibleClassFileVersion(JavaClass javaClass) {
+    int classFileMajorVersion = javaClass.getMajor();
+    // TODO(#343): Java 11 support
+    return 45 <= classFileMajorVersion && classFileMajorVersion <= 52;
   }
 
   private static SymbolReferenceSet scanSymbolReferencesInClass(JavaClass javaClass) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -53,7 +53,9 @@ import org.eclipse.aether.artifact.DefaultArtifact;
 public class StaticLinkageCheckOption {
   
   private static final Options options = configureOptions();
-  
+
+  private static final HelpFormatter helpFormatter = new HelpFormatter();
+
   static CommandLine readCommandLine(String[] arguments) throws ParseException {
     // TODO is this reentrant? Can we reuse it? 
     // https://issues.apache.org/jira/browse/CLI-291
@@ -64,7 +66,6 @@ public class StaticLinkageCheckOption {
       checkInput(commandLine);
       return commandLine;
     } catch (ParseException ex) {
-      HelpFormatter helpFormatter = new HelpFormatter();
       helpFormatter.printHelp("StaticLinkageChecker", options);
       throw ex;
     }
@@ -121,6 +122,7 @@ public class StaticLinkageCheckOption {
               .collect(toImmutableList());
       return jarFilesInArguments;
     } else {
+      helpFormatter.printHelp("StaticLinkageChecker", options);
       throw new ParseException("Missing argument");
     }
   }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -22,6 +22,7 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.logging.Logger;
 import org.apache.maven.repository.internal.MavenRepositorySystemUtils;
 import org.eclipse.aether.DefaultRepositorySystemSession;
@@ -29,6 +30,7 @@ import org.eclipse.aether.RepositorySystem;
 import org.eclipse.aether.RepositorySystemSession;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.ArtifactProperties;
+import org.eclipse.aether.collection.DependencyCollectionContext;
 import org.eclipse.aether.collection.DependencySelector;
 import org.eclipse.aether.connector.basic.BasicRepositoryConnectorFactory;
 import org.eclipse.aether.graph.Dependency;
@@ -84,7 +86,7 @@ public final class RepositoryUtility {
    * customary ~/.m2 directory. If not found, it creates an initially empty repository in
    * a temporary location.
    */
-  public static RepositorySystemSession newSession(RepositorySystem system) {
+  static RepositorySystemSession newSession(RepositorySystem system) {
     DefaultRepositorySystemSession session = createDefaultRepositorySystemSession(system);
     session.setReadOnly();
     return session;
@@ -95,8 +97,29 @@ public final class RepositoryUtility {
    * RepositoryUtility#newSession(RepositorySystem)}, with its dependency selector to include
    * dependencies with 'provided' scope.
    */
-  public static RepositorySystemSession newSessionWithProvidedScope(RepositorySystem system) {
+  static RepositorySystemSession newSessionWithProvidedScope(RepositorySystem system) {
     DefaultRepositorySystemSession session = createDefaultRepositorySystemSession(system);
+
+    // To exclude log4j-api-java9:zip:2.11.1, which is not published.
+    // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/339
+    DependencySelector filteringZipDependencySelector =
+        new DependencySelector() {
+
+          @Override
+          public boolean selectDependency(Dependency dependency) {
+            Artifact artifact = dependency.getArtifact();
+            Map<String, String> properties = artifact.getProperties();
+            // Because StaticLinkageChecker only checks jar file, zip files are not needed
+            logger.fine("Skipping an artifact with type:zip: " + artifact);
+            return !"zip".equals(properties.get("type"));
+          }
+
+          @Override
+          public DependencySelector deriveChildSelector(
+              DependencyCollectionContext dependencyCollectionContext) {
+            return this;
+          }
+        };
 
     // This combination of DependencySelector comes from the default specified in
     // `MavenRepositorySystemUtils.newSession`.
@@ -105,9 +128,10 @@ public final class RepositoryUtility {
         new AndDependencySelector(
             // ScopeDependencySelector takes exclusions. 'Provided' scope is not here to avoid
             // false positive in StaticLinkageChecker.
-            new ScopeDependencySelector(new String[] {"test"}),
+            new ScopeDependencySelector("test"),
             new OptionalDependencySelector(),
-            new ExclusionDependencySelector());
+            new ExclusionDependencySelector(),
+            filteringZipDependencySelector);
     session.setDependencySelector(dependencySelector);
     session.setReadOnly();
 

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java
@@ -28,6 +28,7 @@ import org.eclipse.aether.resolution.DependencyResolutionException;
 import org.junit.Assert;
 import org.junit.Test;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.truth.Truth;
 
 public class DependencyGraphBuilderTest {
@@ -87,6 +88,18 @@ public class DependencyGraphBuilderTest {
     }
     
     Truth.assertThat(coordinates).contains("com.google.code.findbugs:jsr305:jar:3.0.2");
+  }
+
+  @Test
+  public void testGetDirectDependencies_nonExistentZipDependency()
+      throws DependencyCollectionException, DependencyResolutionException {
+    // This artifact depends on log4j-api-java9 (type:zip), which does not exist in Maven central.
+    DefaultArtifact log4j2 = new DefaultArtifact("org.apache.logging.log4j:log4j-api:2.11.1");
+
+    // This should not raise DependencyResolutionException
+    DependencyGraph completeDependencies =
+        DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph(ImmutableList.of(log4j2));
+    Truth.assertThat(completeDependencies.list()).isNotEmpty();
   }
 
   @Test

